### PR TITLE
Fixing generator, third attempt.

### PR DESCRIPTION
**Pull request by mithro**
This needs https://github.com/web-animations/web-animations-js/pull/293 to be merged before this one.

This uses code from the first fixing attempt but instead uses direct function calls rather than postMessage. It also fixed the "load" button to work (which never worked in the original fixing attempt).

### CODE REVIEW

File: test/test-generator.html [hunk 1]
```diff
@@ -84,32 +84,63 @@
   this.properties = properties;
 }
 
-function loadFile(){
+function getEndTime(timeline) {
+  var players = timeline.getCurrentPlayers();
+
+  var endTime = 0;
+  players.forEach(function(player) {
+    endTime = Math.max(endTime, player.source.iterationDuration + player.source.specified.startDelay);
```
Comment (dstoc) [inline test/test-generator.html]: I don't think this takes iterations into account, try player.source.endTime

File: test/test-generator.html [hunk 2]
```diff
@@ -84,32 +84,63 @@
...
+
   var file = document.getElementById("iframeSrc").value;
-  document.getElementById("frame").src = file;
+
+  var iframe = document.getElementById("frame");
+  iframe.onload = function() {
+    iframe.onload = callback;
+    iframe.src = file + '#' + mode;
+  };
+  // Set to null to cause a reload when we change below.
```
Comment (dstoc) [inline test/test-generator.html]: I'm not sure what "below" is in reference to.

File: test/test-generator.html [hunk 3]
```diff
@@ -84,32 +84,63 @@
...
-    process(obProp);
-    testLength = 0;
-    window.frames[0].animTestRunner.players.forEach(function(player) {
-      testLength = Math.max(testLength, player.source.animationDuration + player.source.specified.startDelay);
-    });
-    pause();
-    setTestCurrentTime(0);
-    setTimeout(generate, frameTime);
-  }, 100);
+function go() {
```
Comment (dstoc) [inline test/test-generator.html]: go -> startGeneratingExpectations ?

File: test/test-generator.html [hunk 4]
```diff
@@ -84,32 +84,63 @@
...
+      generate(t);
+      t += interval;
+    }
+
+    output.innerHTML += '}, "Auto generated tests");\n';
+  });
 }
 
-function process(rawString){
+function findObjectsToCheck(rawString){
```
Comment (dstoc) [inline test/test-generator.html]: Not sure what "Object" means here, is it an Element?

File: test/test-generator.html [hunk 5]
```diff
@@ -160,47 +193,39 @@
...
-  });
-  testCurrentTime = time;
-}
+  // If all the properties are the same for every object, just pass the list.
+  var allsame = properties.every(function(x) { return x === properties[0] });
+  if (allsame) {
+    newCheck = newCheck.substring(0, newCheck.length-2);
+    newCheck += properties[0] + ");\n";
 
+  // Otherwise give a list with each
```
Comment (dstoc) [inline test/test-generator.html]: This comment probably belongs inside the else branch.

File: test/test-generator.html (hunk 1 above)
Comment (mithro) [inline test/test-generator.html]: Fixed.

File: test/test-generator.html (hunk 2 above)
Comment (mithro) [inline test/test-generator.html]: Fixed. The code use to be below.

File: test/test-generator.html (hunk 3 above)
Comment (mithro) [inline test/test-generator.html]: Fixed.

File: test/test-generator.html (hunk 4 above)
Comment (mithro) [inline test/test-generator.html]: Fixed.

File: test/test-generator.html (hunk 5 above)
Comment (mithro) [inline test/test-generator.html]: Fixed.

File: test/test-generator.html [hunk 6]
```diff
@@ -160,47 +194,40 @@
...
-function setTestCurrentTime(time){
-  window.frames[0].animTestRunner.players.forEach(function(player) {
-    player.currentTime = time;
-  });
-  testCurrentTime = time;
-}
+  // If all the properties are the same for every object, just pass the list.
+  var allsame = properties.every(function(x) { return x === properties[0] });
+  if (allsame) {
+    newCheck = newCheck.substring(0, newCheck.length-2);
```
Comment (dstoc) [inline test/test-generator.html]: Can we just add "[\n" in the else to avoid having to remove it here?

File: test/test-generator.html [hunk 7]
```diff
@@ -160,47 +194,40 @@
...
+    newCheck += properties[0] + ");\n";
 
+  } else {
+    // Otherwise give a list with each
+
+    for(var j in properties) {
```
Comment (dstoc) [inline test/test-generator.html]: Space after for

File: test/test-generator.html [hunk 8]
```diff
@@ -84,32 +84,64 @@
...
+
+    output.innerHTML += "timing_test(function() {\n";
+    var testWindow = document.getElementById("frame").contentWindow;
+    var interval = parseFloat(document.getElementById("interval").value);
+
+    var endTime = Infinity;
+
+    var t = 0;
+    while(t <= endTime) {
+      testWindow.testharness_timeline.setTime(t*1000.0);
```
Comment (dstoc) [inline test/test-generator.html]: spaces around *

File: test/test-generator.html (hunk 6 above)
Comment (dstoc) [inline test/test-generator.html]: spaces around -

File: test/test-generator.html [hunk 9]
```diff
@@ -84,32 +84,64 @@
   this.properties = properties;
 }
 
-function loadFile(){
+function getEndTime(timeline) {
+  var players = timeline.getCurrentPlayers();
+
+  var endTime = 0;
```
Comment (dstoc) [inline test/test-generator.html:90]: This is perhaps  a little more idiomatic for JavaScript:
var endTime = players.reduce(function(a, b) { return Math.max(a.source.endTime, b.source.endTime); }, 0);

**Comment by dstoc** [comment]
lgtm, just a few more small comments

File: test/test-generator.html (hunk 7 above)
Comment (mithro) [inline test/test-generator.html]: Fixed.

File: test/test-generator.html (hunk 6 above)
Comment (mithro) [inline test/test-generator.html]: Fixed both. I think this got lost in the rebase...

File: test/test-generator.html (hunk 8 above)
Comment (mithro) [inline test/test-generator.html]: Fixed.

**Comment by mithro** [comment]
Poke?

**Pull request merged**